nexus/mgs-updates/src/mgs_clients.rs

```diff
@@ -6,13 +6,64 @@
 
 use futures::Future;
 use gateway_client::Client;
+use gateway_client::HostPhase1HashError;
+use gateway_client::types::SpComponentResetError;
 use slog::Logger;
 use slog::warn;
 use std::collections::VecDeque;
+use std::fmt;
 use std::sync::Arc;
 
 pub(super) type GatewayClientError =
     gateway_client::Error<gateway_client::types::Error>;
+pub(super) type GatewaySpComponentResetError =
+    gateway_client::Error<SpComponentResetError>;
+
+/// Helper trait for error types we can get from various gateway_client API
+/// calls.
+///
+/// For each error type, we need to know whether we should retry the request (by
+/// sending it to the next MGS instance).
+pub(super) trait RetryableMgsError: fmt::Display {
+    fn should_try_next_mgs(&self) -> bool;
```

Comment:
Take it or leave it: I wonder if this ought to be in `gateway-client`? There's other code in Nexus that has similar behavior of attempting a request through both MGSes that could also use this...

Reply:
I can see that, although I worry slightly that this is already a little update-specific. If it were in `gateway-client` I'd be tempted to call it `err.is_retryable()`, except the real driver for this at all is to be able to say "if you get back `SpComponentResetError::ResetSpOfLocalSled`, you must send the request to the other MGS". So it's not really "retryable" so much as "you might have success if you tried the other MGS instead of this one". Maybe the name should reflect that more precisely, and then it would be fine to put it in gateway-client?